data/transactions/logic/eval.go

```diff
@@ -0,0 +1,709 @@
+		value = value | (uint64(b) & 0x0ff)
+	}
+	cx.stack[last].Uint = value
+	cx.stack[last].Bytes = nil
+}
+
+func opBitOr(cx *evalContext) {
+	last := len(cx.stack) - 1
+	prev := last - 1
+	cx.stack[prev].Uint = cx.stack[prev].Uint | cx.stack[last].Uint
+	cx.stack = cx.stack[:last]
+}
+
+func opBitAnd(cx *evalContext) {
+	last := len(cx.stack) - 1
+	prev := last - 1
+	cx.stack[prev].Uint = cx.stack[prev].Uint & cx.stack[last].Uint
+	cx.stack = cx.stack[:last]
+}
+
+func opBitXor(cx *evalContext) {
+	last := len(cx.stack) - 1
+	prev := last - 1
+	cx.stack[prev].Uint = cx.stack[prev].Uint ^ cx.stack[last].Uint
+	cx.stack = cx.stack[:last]
+}
+
+func opBitNot(cx *evalContext) {
+	last := len(cx.stack) - 1
+	cx.stack[last].Uint = cx.stack[last].Uint ^ 0xffffffffffffffff
+}
+
+func opIntConstBlock(cx *evalContext) {
+	cx.intc, cx.nextpc, cx.err = parseIntcblock(cx.program, cx.pc)
+}
+
+func opIntConstN(cx *evalContext, n uint) {
+	if n >= uint(len(cx.intc)) {
+		cx.err = fmt.Errorf("intc [%d] beyond %d constants", n, len(cx.intc))
+		return
+	}
+	cx.stack = append(cx.stack, stackValue{Uint: cx.intc[n]})
+}
+func opIntConstLoad(cx *evalContext) {
+	n := uint(cx.program[cx.pc+1])
+	opIntConstN(cx, n)
+	cx.nextpc = cx.pc + 2
+}
+func opIntConst0(cx *evalContext) {
+	opIntConstN(cx, 0)
+}
+func opIntConst1(cx *evalContext) {
+	opIntConstN(cx, 1)
+}
+func opIntConst2(cx *evalContext) {
+	opIntConstN(cx, 2)
+}
+func opIntConst3(cx *evalContext) {
+	opIntConstN(cx, 3)
+}
+
+func opByteConstBlock(cx *evalContext) {
+	cx.bytec, cx.nextpc, cx.err = parseBytecBlock(cx.program, cx.pc)
+}
+func opByteConstN(cx *evalContext, n uint) {
+	if n >= uint(len(cx.bytec)) {
+		cx.err = fmt.Errorf("bytec [%d] beyond %d constants", n, len(cx.bytec))
+		return
+	}
+	cx.stack = append(cx.stack, stackValue{Bytes: cx.bytec[n]})
+}
+func opByteConstLoad(cx *evalContext) {
+	n := uint(cx.program[cx.pc+1])
+	opByteConstN(cx, n)
+	cx.nextpc = cx.pc + 2
+}
+func opByteConst0(cx *evalContext) {
+	opByteConstN(cx, 0)
+}
+func opByteConst1(cx *evalContext) {
+	opByteConstN(cx, 1)
+}
+func opByteConst2(cx *evalContext) {
+	opByteConstN(cx, 2)
+}
+func opByteConst3(cx *evalContext) {
+	opByteConstN(cx, 3)
+}
+
+func opArgN(cx *evalContext, n uint64) {
+	if n >= uint64(len(cx.Txn.Lsig.Args)) {
+		cx.err = fmt.Errorf("cannot load arg[%d] of %d", n, len(cx.Txn.Lsig.Args))
+		return
+	}
+	cx.stack = append(cx.stack, stackValue{Bytes: cx.Txn.Lsig.Args[n]})
+}
+func opArg(cx *evalContext) {
+	n := uint64(cx.program[cx.pc+1])
+	opArgN(cx, n)
+	cx.nextpc = cx.pc + 2
+}
+func opArg0(cx *evalContext) {
+	opArgN(cx, 0)
+}
+func opArg1(cx *evalContext) {
+	opArgN(cx, 1)
+}
+func opArg2(cx *evalContext) {
+	opArgN(cx, 2)
+}
+func opArg3(cx *evalContext) {
+	opArgN(cx, 3)
+}
+
+func opBnz(cx *evalContext) {
+	last := len(cx.stack) - 1
+	cx.nextpc = cx.pc + 2
+	isNonZero := cx.stack[last].Uint != 0
+	cx.stack = cx.stack[:last] // pop
+	if isNonZero {
+		cx.nextpc += int(cx.program[cx.pc+1])
```

Comment:
is that enough ? i.e. you're limiting the jump to 256 bytes. giving the the program could be 1K, that's quite a limitation.

Reply:
The longest program I've written has 40 bytes of actual content. If we extend that to 2 bytes and limit it to 32767 so that a future language extension could allow for _signed_ jumps, that'd probably good enough for a long while. Ok. probably add this tomorrow.